### PR TITLE
Add include_path option for dask.bag.read_text

### PR DESCRIPTION
**Pull request by gyf304**
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by jakirkham** [comment]
@mrocklin, would you have time to review this or recommend someone who could?

**Comment by gyf304** [comment]
I'm internally debating if `include_path` should return a tuple of `(bag, paths)` like `dask.bytes` does, or return a bag of `(line, path)` like what `dask.dataframe` does. Not sure which way is more useful.

**Comment by TomAugspurger** [comment]
@gyf304 just FYI, when you force push to the branch we don't get notifications and it makes things a bit harder to review. Can you just push additional commits? We squash on merge.

> internally debating if include_path should return a tuple of (bag, paths) like dask.bytes does, or return a bag of (line, path) like what dask.dataframe does.

Can you include documentation / examples for what this does? That may aid in determining what is more natural. My initial reaction is that a keyword probably shouldn't change the return type of a method (from `Bag` to `Tuple`), though it's hard to say.

**Comment by gyf304** [comment]
Sorry about the force push, was used to the workflow of gerrit of doing `git commit --amend`. Will directly commit next time.

Example for first approach
```python
> bag, paths = dask.bag.read_text("./test/*.txt", include_path=True)
> paths
["/home/.../test/1.txt", "/home/.../test/2.txt"]
> len(paths) == bag.npartitions
True
```

Example for second approach
```python
> bag = dask.bag.read_text("./test/*.txt", include_path=True)
> bag.take(1)
(("hello", "/home/.../test/1.txt"),)
```

**Comment by TomAugspurger** [comment]
Do you have a suggestion for which is more useful?

I would think that the first is relatively easily accomplished by

```python
paths = glob.glob("*/test/*.txt")
bag = db.read_text(paths)
```

The second is harder to achieve without the `include_path` keyword.

**Comment by gyf304** [comment]
I agree. This also mirrors well with the `include_path_column` option in `dask.dataframe.read_csv`
https://docs.dask.org/en/latest/dataframe-api.html#dask.dataframe.read_csv
I already updated the patch to the second behavior after the same thought process.

File: dask/bag/tests/test_text.py [hunk 1]
```diff
@@ -77,6 +93,18 @@ def test_files_per_partition():
 
             assert b.count().compute() == 20, "All 20 lines should be read"
 
+            with pytest.warns(UserWarning):
```
Comment (TomAugspurger) [inline dask/bag/tests/test_text.py]: Why are we asserting a warning here?

File: dask/bag/tests/test_text.py [hunk 2]
```diff
@@ -48,9 +53,20 @@ def test_read_text(fmt, bs, encoding):
...
+            compression=fmt,
+            blocksize=bs,
+            encoding=encoding,
+            include_path=include_path,
+        )
+        b = o.pluck(0) if include_path else o
         (L,) = compute(b)
         assert "".join(L) == expected
+        if include_path:
+            (paths,) = compute(o.pluck(1))
```
Comment (TomAugspurger) [inline dask/bag/tests/test_text.py]: Can we be a bit stronger about this assertion? Some kind of assertion that all of the lines within a partition have the expected path?

File: dask/bag/text.py [hunk 3]
```diff
@@ -63,9 +68,14 @@ def read_text(
 
     >>> b = read_text('largefile.txt', blocksize='10MB')  # doctest: +SKIP
 
+    Get file paths of the bag by setting include_path=True
+
+    >>> b = read_text('myfiles.*.txt', include_path=True) # doctest: +SKIP
```
Comment (TomAugspurger) [inline dask/bag/text.py]: Can you show a `.pluck` to show an example of what the structure will looklike?

File: dask/bag/text.py [hunk 4]
```diff
@@ -119,6 +150,17 @@ def file_to_blocks(lazy_file):
             yield line
 
 
+def file_to_blocks_with_path(lazy_file):
```
Comment (TomAugspurger) [inline dask/bag/text.py]: Slight preference to update `file_to_blocks` to take a boolean `include_path` and yield `(line, lazy_file.path)` if so.

File: dask/bag/tests/test_text.py (hunk 1 above)
Comment (gyf304) [reply dask/bag/tests/test_text.py]: This is from line 72(master) / 88(HEAD). This was already present in the old test.

File: dask/bag/text.py [hunk 5]
```diff
@@ -63,9 +68,16 @@ def read_text(
 
     >>> b = read_text('largefile.txt', blocksize='10MB')  # doctest: +SKIP
 
+    Get file paths of the bag by setting include_path=True
+
+    >>> b = read_text('myfiles.*.txt', include_path=True) # doctest: +SKIP
+    >>> b.pluck(1).take(1) # doctest: +SKIP
+    ('.../myfiles.0.txt',)
```
Comment (TomAugspurger) [inline dask/bag/text.py]: I think it'd be a bit easier to understand with something like

```
>>> b.pluck(1)
(('first line of the first file\n', 'myfiles.0.txt'),)
```

Why is the `...` present in your output? Are the paths relative or absolute?

Comment (gyf304) [reply dask/bag/text.py]: Think it should be
```
>>> b.take(1)
(('first line of the first file\n', '.../myfiles.0.txt'),)
```
Paths are absolute. The `...` is just ellipsis for path. Open to suggestions for alternatives to `...`

Comment (gyf304) [reply dask/bag/text.py]: Maybe something generic like `/home/dask/myfiles.0.txt`?

**Comment by TomAugspurger** [comment]
Thanks @gyf304!

**Comment by SultanOrazbayev** [comment]
This is great, thanks @gyf304!

**Pull request merged**